mythril/analysis/modules/state_change_external_calls.py

```diff
@@ -0,0 +1,197 @@
+
+    def __init__(self):
+        """"""
+        super().__init__(
+            name="State Change After External calls",
+            swc_id=REENTRANCY,
+            description=DESCRIPTION,
+            entrypoint="callback",
+            pre_hooks=[
+                "CALL",
+                "SSTORE",
+                "DELEGATECALL",
+                "STATICCALL",
+                "CREATE",
+                "CALLCODE",
+            ],
+        )
+
+    def execute(self, state: GlobalState):
+        """
+
+        :param state:
+        :return:
+        """
+        self._issues.extend(self._analyze_state(state))
+        return self.issues
+
+    @staticmethod
+    def _add_external_call(
+        state: GlobalState, annotations: List[StateChangeCallsAnnotation]
+    ) -> None:
+        call = get_call_from_state(state)
+        gas = state.mstate.stack[-1]
+        to = state.mstate.stack[-2]
+        if call is None:
+            return
+        try:
+            constraints = copy(state.node.constraints)
+            solver.get_model(
+                constraints + [UGT(gas, symbol_factory.BitVecVal(2300, 256))]
+            )
+
+            # Check whether we can also set the callee address
+            try:
+                constraints += [to == 0xDEADBEEFDEADBEEFDEADBEEFDEADBEEFDEADBEEF]
+                solver.get_model(constraints)
+                annotations[0].calls.append(
+                    CallIssue(call=call, user_defined_address=True)
+                )
+            except UnsatError:
+                annotations[0].calls.append(
+                    CallIssue(call=call, user_defined_address=False)
+                )
+
+        except UnsatError:
+            pass
+
+    @staticmethod
+    def _analyze_state(state: GlobalState) -> List[Issue]:
+        """
+
+        :param state:
+        :return:
+        """
+        address = state.get_current_instruction()["address"]
+        annotations = cast(
+            List[StateChangeCallsAnnotation],
+            list(state.get_annotations(StateChangeCallsAnnotation)),
+        )
+        if len(annotations) == 0:
+            log.debug("Creating annotation for state")
+            state.annotate(StateChangeCallsAnnotation())
+            annotations = cast(
+                List[StateChangeCallsAnnotation],
+                list(state.get_annotations(StateChangeCallsAnnotation)),
+            )
+        opcode = state.get_current_instruction()["opcode"]
+
+        if opcode in ("SSTORE", "CREATE", "CREATE2"):
+            return StateChange._handle_state_change(
+                state, address=address, annotation=annotations[0]
+            )
+        call = get_call_from_state(state)
+
+        if call is None:
+            return []
+
+        if StateChange._balance_change(call.value):
+            return StateChange._handle_state_change(
+                state, address=address, annotation=annotations[0]
+            )
+
+        if opcode == "CALL":
```

Comment:
why are only calls considered here?

Reply:
This is to detect state change after sending gas with call.value() right?

Reply:
This isn't to detect state changes, this is the check for a call. So I'm wondering if other types of calls should be considered here as well.

Reply:
Argh! nice catch, some residual if condition.

Reply:
Oh! I am not a security expert, do reentrancy sort of problems happen with other calls?